libs/knowledge-store/ragstack_knowledge_store/langchain/cassandra.py

```diff
@@ -0,0 +1,198 @@
+
+    async def aembed_query(self, text: str) -> List[float]:
+        return await self.embeddings.aembed_query(text)
+
+
+def _row_to_document(row) -> Document:
+    return Document(
+        page_content=row.text_content,
+        metadata={
+            knowledge_store.CONTENT_ID: row.content_id,
+            "kind": row.kind,
+        },
+    )
+
+
+def _results_to_documents(results: Optional[ResponseFuture]) -> Iterable[Document]:
+    if results:
+        for row in results:
+            yield _row_to_document(row)
+
+
+class CassandraKnowledgeStore(KnowledgeStore):
+    def __init__(
+        self,
+        embedding: Embeddings,
+        edge_extractors: List[EdgeExtractor],
+        *,
+        node_table: str = "knowledge_nodes",
+        edge_table: str = "knowledge_edges",
+        session: Optional[Session] = None,
+        keyspace: Optional[str] = None,
+        setup_mode: SetupMode = SetupMode.SYNC,
+        concurrency: int = 20,
+    ):
+        """
+        Create the hybrid knowledge store.
+        Parameters configure the ways that edges should be added between
+        documents. Many take `Union[bool, Set[str]]`, with `False` disabling
+        inference, `True` enabling it globally between all documents, and a set
+        of metadata fields defining a scope in which to enable it. Specifically,
+        passing a set of metadata fields such as `source` only links documents
+        with the same `source` metadata value.
+        Args:
+            embedding: The embeddings to use for the document content.
+            edge_extractors: Edge extractors to use for linking knowledge chunks.
+            concurrency: Maximum number of queries to have concurrently executing.
+            setup_mode: Mode used to create the Cassandra table (SYNC,
+                ASYNC or OFF).
+        """
+        self._embedding = embedding
+        self._session = check_resolve_session(session)
+        _keyspace = check_resolve_keyspace(keyspace)
+        _setup_mode = getattr(knowledge_store.SetupMode, setup_mode.name)
+
+        self.store = knowledge_store.KnowledgeStore(
+            embedding=EmbeddingModelAdapter(embedding),
+            edge_extractors=edge_extractors,
+            node_table=node_table,
+            edge_table=edge_table,
+            session=self._session,
+            keyspace=_keyspace,
+            setup_mode=_setup_mode,
+            concurrency=concurrency,
+        )
+
+        self._query_by_embedding = self._session.prepare(
+            f"""
+            SELECT content_id, kind, text_content
+            FROM {_keyspace}.{node_table}
+            ORDER BY text_embedding ANN OF ?
+            LIMIT ?
+            """
+        )
+
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
+        return self._embedding
+
+    def add_nodes(
+        self,
+        nodes: Iterable[Node] = None,
+        **kwargs: Any,
+    ):
+        _nodes = []
+        for node in nodes:
```

Comment:
Why does this need code? It would probably make more sense to just call `self.store.add_nodes(nodes)` and let that handle the kinds of nodes that are supported or not.

Reply:
This code makes the adaptation from LangChain's `Node` to RagStack's `Node`.
It is required if we don't want one layer to leak into the other.

Reply:
Ah. That makes sense. I guess the other option would be for us to hide the non-text nodes from LangChain (eg., not have a `Node` or `add_nodes` yet). That may reduce the apparent size of the interface -- we could still say the underlying thing has support for multi-modal nodes, and come back and add that once it's in LangChain.

Either way seems OK though if we think the additional node isn't going to add complexity.